File: php/spec/elements/empty.spec.php
<?php declare(strict_types=1);

use Eno\{EmptyElement, Section};

describe('EmptyElement', function() {
  beforeAll(function() {
    $this->_context = (object) [];
    $this->instruction = (object) [ 'key' => 'language' ];
    $this->section_instruction = (object) [
      'depth' => 0,
      'key' => 'mock',
      'subinstructions' => []
    ];
    $this->parent = new Section($this->_context, $this->section_instruction);
  });

  beforeEach(function() {
    $this->empty = new EmptyElement($this->_context, $this->instruction, $this->parent);
  });

  it('is untouched after initialization', function() {
    expect($this->empty->touched)->toBe(false);
  });

  describe('raw()', function() {
    it('returns a native object representation', function() {
      expect($this->empty->raw())->toEqual([ 'language' => null ]);
    });
  });

  describe('__toString()', function() {
    it('returns a debug representation', function() {
      expect((string)$this->empty)->toEqual('[EmptyElement key="language"]');
    });
  });

  describe('touch()', function() {
    it('touches the element', function() {
      $this->empty->touch();
      expect($this->empty->touched)->toBe(true);
    });
  });

  describe('value()', function() {
    beforeEach(function() {
      $this->value = $this->empty->value();
    });

    it('returns null', function() {
      expect($this->value)->toBe(null);
    });

    it('touches the element', function() {
      expect($this->empty->touched)->toBe(true);
    });
  });
});

File: ruby/spec/elements/empty.spec.rb
# frozen_string_literal: true

describe Enolib::Empty do
  before(:each) do
    @empty = Enolib.parse('element:').empty
  end

  it 'is untouched after initialization' do
    expect(@empty.instance_variable_defined?(:@touched)).to be false
  end

  describe '#error' do
    context 'with a message' do
      it 'returns a custom error' do
        expect(@empty.error('my custom error').message).to match_snapshot
      end
    end

    context 'with a message Proc' do
      it 'returns a custom error' do
        message_proc = ->(element) do
          "my custom generated message for empty element '#{element.string_key}'"
        end

        expect(@empty.error(message_proc).message).to match_snapshot
      end
    end
  end

  describe '#raw' do
    it 'returns a native object representation' do
      expect(@empty.raw).to eq({
          key: 'element',
          type: :empty_element
        })
    end
  end

  describe '#touch' do
    it 'touches the element' do
      @empty.touch
      expect(@empty.instance_variable_defined?(:@touched)).to be true
    end
  end

  describe '#to_s' do
    it 'returns a debug abstraction' do
      expect(@empty.to_s).to eq('#<Enolib::Empty key=element>')
    end
  end
end
